### PR TITLE
Dump IO Usage

### PR DESCRIPTION
**Pull request by pratikpatil024**
[Jira Task: POS-573](https://polygon.atlassian.net/browse/POS-573)

### CODE REVIEW

File: core/state/statedb.go [hunk 1]
```diff
@@ -349,6 +350,34 @@ func (sw *StateDB) ApplyMVWriteSet(writes []blockstm.WriteDescriptor) {
...
+		// fmt.SprintF is expensive, but can be sued for debugging
+		str += fmt.Sprintf("%v", s.txIndex) + ", " + fmt.Sprintf("%v", s.incarnation) + ", " + fmt.Sprintf("%v", val.V.TxnIndex) + ", " + fmt.Sprintf("%v", val.V.Incarnation) + ", " + hex.EncodeToString(val.Path) + ", " + "Read\n"
+	}
+
+	return str
+}
+
+// get writeMap Dump of format: "TxIdx, Inc, Path, Write"
+func (s *StateDB) GetWriteMapDump() string {
+	str := ""
```
Comment (JekaMas) [inline core/state/statedb.go:366]: Do we need this API fast or it is not critical? If it should be performant, then
1. remove all fmt.Sprintf in favor of 1 fmt.Sprintf with 1 format string
2. try string.Builder
3. maybe it worth it to stick with []byte rather then strings

**Review by JekaMas — APPROVED** [review]

File: core/state/statedb.go (hunk 1 above)
Comment (pratikpatil024) [reply core/state/statedb.go:366]: Right, so we will be using this only to debug, that was the reason I went ahead with fmt.Sprintf.
And about using []byte, we wanted to make the result readable, therefore using strings.

Comment (JekaMas) [reply core/state/statedb.go:366]: However it's possible to convert []byte as a very last step before sending data to a user.

Comment (pratikpatil024) [reply core/state/statedb.go:366]: Ohh, okay. I will do that change. Thank you.

File: core/state/statedb.go [hunk 2]
```diff
@@ -349,6 +349,59 @@ func (sw *StateDB) ApplyMVWriteSet(writes []blockstm.WriteDescriptor) {
...
+	TxInc  int
+	VerIdx int
+	VerInc int
+	Path   []byte
+	Op     string
+}
+
+// get readMap Dump of format: "TxIdx, Inc, Path, Read"
+func (s *StateDB) GetReadMapDump() []DumpStruct {
+	res := []DumpStruct{}
```
Comment (JekaMas) [inline core/state/statedb.go:363]: could we preallocate the slice here? like
```
readList := s.MVReadList()
res := make([]DumpStruct, 0, len(readList))
for _, val := range readList {
// the rest of the code
}
```

File: core/state/statedb.go [hunk 3]
```diff
@@ -349,6 +349,59 @@ func (sw *StateDB) ApplyMVWriteSet(writes []blockstm.WriteDescriptor) {
...
+		}
+		res = append(res, *temp)
+	}
+
+	return res
+}
+
+// get writeMap Dump of format: "TxIdx, Inc, Path, Write"
+func (s *StateDB) GetWriteMapDump() []DumpStruct {
+	res := []DumpStruct{}
```
Comment (JekaMas) [inline core/state/statedb.go:382]: the same thing with preallocation

File: eth/tracers/api.go [hunk 4]
```diff
@@ -615,25 +632,80 @@ func (api *API) traceBlock(ctx context.Context, block *types.Block, config *Trac
...
+	RWstruct := []state.DumpStruct{}
 	// Feed the transactions into the tracers and return
 	var failed error
 	blockCtx := core.NewEVMBlockContext(block.Header(), api.chainContext(ctx), nil)
+
+	london := api.backend.ChainConfig().IsLondon(block.Number())
+
 	for i, tx := range txs {
+		// copy of statedb
+		statedb := statedb.Copy()
```
Comment (JekaMas) [inline eth/tracers/api.go:646]: could we make this changes optional? to not influence original API and performance? maybe add a flag to the request or something?

File: eth/tracers/api.go [hunk 5]
```diff
@@ -615,25 +632,80 @@ func (api *API) traceBlock(ctx context.Context, block *types.Block, config *Trac
...
+
+		structRead := statedb.GetReadMapDump()
+		structWrite := statedb.GetWriteMapDump()
+
+		RWstruct = append(RWstruct, structRead...)
+		RWstruct = append(RWstruct, structWrite...)
+	}
+
+	for _, val := range RWstruct {
+		IOdump += fmt.Sprintf("%v", val.TxIdx) + ", " + fmt.Sprintf("%v", val.TxInc) + ", " + fmt.Sprintf("%v", val.VerIdx) + ", " + fmt.Sprintf("%v", val.VerInc) + ", " + hex.EncodeToString(val.Path) + ", " + val.Op
```
Comment (JekaMas) [inline eth/tracers/api.go:700]: Could we have single format siring?

File: eth/tracers/api.go [hunk 6]
```diff
@@ -615,25 +632,80 @@ func (api *API) traceBlock(ctx context.Context, block *types.Block, config *Trac
...
 	}
+
+	// make sure that the file exists and write IOdump
+	err = ioutil.WriteFile(filepath.Join(path, "data.csv"), []byte(fmt.Sprint(IOdump)), 0600)
+	if err != nil {
+		fmt.Println("Writing IOdump to the file FAILED!\n IOdump:", IOdump)
```
Comment (JekaMas) [inline eth/tracers/api.go:706]: Shall we return an error here?

**Comment by cffls** [comment]
Looks like linter failed. Could you fix before merging?

**Review by cffls — APPROVED** [review]

**Comment by pratikpatil024** [comment]
@cffls @JekaMas I am not getting the lint errors on my local machine. Also, these lines, where the CI is throwing an error is not present in the code.

```
Error: core/parallel_state_processor.go:137:1: cognitive complexity 32 of func `(*ParallelStateProcessor).Process` is high (> 30) (gocognit)
func (p *ParallelStateProcessor) Process(block *types.Block, statedb *state.StateDB, cfg vm.Config) (types.Receipts, []*types.Log, uint64, error) {
^
Error: core/parallel_state_processor.go:101:3: if statements should only be cuddled with assignments (wsl)
		if _, ok := task.statedb.MVReadMap()[string(task.result.BurntContractAddress.Bytes())]; ok {
		^
Error: core/parallel_state_processor.go:195:4: branch statements should not be cuddled if block has more than two lines (wsl)
			break
```

**Comment by pratikpatil024** [comment]
> @cffls @JekaMas I am not getting the lint errors on my local machine. Also, these lines, where the CI is throwing an error is not present in the code.
> 
> ```
> Error: core/parallel_state_processor.go:137:1: cognitive complexity 32 of func `(*ParallelStateProcessor).Process` is high (> 30) (gocognit)
> func (p *ParallelStateProcessor) Process(block *types.Block, statedb *state.StateDB, cfg vm.Config) (types.Receipts, []*types.Log, uint64, error) {
> ^
> Error: core/parallel_state_processor.go:101:3: if statements should only be cuddled with assignments (wsl)
> 		if _, ok := task.statedb.MVReadMap()[string(task.result.BurntContractAddress.Bytes())]; ok {
> 		^
> Error: core/parallel_state_processor.go:195:4: branch statements should not be cuddled if block has more than two lines (wsl)
> 			break
> ```

Actually these changes were added in this [commit](https://github.com/maticnetwork/bor/commit/98f8b19bc5c5b4690cf66655bfb1e7e423f8f961) in the base repo (`block-atm`). And I think because of that the linters are failing. These will be resolved, but once lint error which might need some thinking is. this one: 

```
Error: core/parallel_state_processor.go:142:1: cognitive complexity 32 of func `(*ParallelStateProcessor).Process` is high (> 30) (gocognit)
```

**Comment by cffls** [comment]
I guess it is okay to merge then, since some failed linters are caused by another commit. We can fix linter issues altogether in this PR: https://github.com/maticnetwork/bor/pull/490

**Pull request merged**